frame/staking/src/lib.rs

```diff
@@ -819,8 +896,49 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
-			Self::ensure_storage_upgraded();
+		fn on_initialize(now: T::BlockNumber) {
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::election_status() == OffchainElectionStatus::None &&
+				// and an era is about to be changed.
+				Self::is_current_session_final()
+			{
+				// TODO: maybe we can be naive like im-online and just assume block == slot?
+				let next_session_change =
+					T::NextSessionChange::predict_next_session_change(now);
+				if let Some(remaining) = next_session_change.checked_sub(&now) {
+					if remaining <= T::ElectionLookahead::get() && !remaining.is_zero() {
+						// Set the flag to make sure we don't waste any compute here in the same era
+						// after we have triggered the offline compute.
+						<ElectionStatus<T>>::put(
+							OffchainElectionStatus::<T::BlockNumber>::Triggered(now)
+						);
+						frame_support::print("detected a good block to trigger offchain worker.");
+					}
+				} else {
+					frame_support::print("predicted next authority set change to be in the past.");
+				}
+			}
+		}
+
+		fn offchain_worker(now: T::BlockNumber) {
+			// TODO: add runtime logging.
+			if sp_io::offchain::is_validator() {
+				if Self::election_status() == OffchainElectionStatus::<T::BlockNumber>::Triggered(now) {
```

Comment:
This will run multiple times, since the `election_status()` is only changed on-chain, so after the transaction is actually included. Instead local offchain worker should manage it's state in `Local Storage`.

Reply:
Is it though? the `if` is also checking `now` to be equal to the contained value in `Triggered` (now renamed to `Open(now)`). So it should be fine? And since the same flag is used to see when normal accounts can submit a solution, I think it should be on-chain.

Reply:
I added a new offchain `check_set_flag` thingy.